### PR TITLE
feat(ipam): add ipam ip data source

### LINKED ISSUE #2075: Expose instances' private network IP addresses in attributes and ideally as writable values
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

When creating a new `scaleway_instance_server` attached to a private network, there is no ability to configure or retrieve the IP address dynamically allocated by Scaleway's IPAM to said instance.

That data is available in the console, however.

It would be helpful to get a of said IP address in the terraform state; this might for instance be shown on the `scaleway_instance_private_nic`, and/or on the instance's `private_network`.

Once the backend supports it, also enable users to set their preferred IP address when declaring those resources: `scaleway_instance_server.private_network.ipv6_address` or `scaleway_instance_private_nic.ipv6_address`.

This is also required in order to have security groups where servers are only allowed to communicate to a subset of servers within a private network.

### New or Affected Resource(s)

* `scaleway_instance_server`
* `scaleway_instance_private_nic`
* `scaleway_vpc_private_network`

### Potential Terraform Configuration

```hcl
resource "scaleway_vpc" "vpc" {
  name = "vpc"
}

resource "scaleway_vpc_private_network" "pn" {
  name = "pn"
  ipv4_subnet {
    subnet = "172.16.10.0/24"
  }
  ipv6_subnet {
    subnet = "fd53:0:1:2::/64"
  }
}

resource "scaleway_instance_server" "server" {
  type = "..."
  image = "..."
  name = "server"
  enable_ipv6 = true
  private_network {
    pn_id = scaleway_vpc_private_network.pn.id
    # Ideally, allow us to set the values according to our internal network layout here
    ipv4_address = "172.16.10.4/24"
    ipv6_address = "fd53:0:1:2::4"
  }
}

output "server_private_ipv6" {
  value = module.scaleway_instance_server.server.private_network.ipv6_address
}
```

### PR DESCRIPTION
**Pull request by Codelax**
Fix #2075

### CODE REVIEW

File: scaleway/data_source_ipam_ip.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+				Optional: true,
+			},
+			"resource_id": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"mac_address": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
```
Comment (quantumsheep) [inline scaleway/data_source_ipam_ip.go:30]: Maybe require one of these fields, what do you think?

File: scaleway/data_source_ipam_ip.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+		Region:           region,
+		OrganizationID:   nil,
+		PrivateNetworkID: expandStringPtr(d.Get("private_network_id")),
+		SubnetID:         nil,
+		Attached:         nil,
+		ResourceID:       expandStringPtr(d.Get("resource_id")),
+		ResourceType:     ipam.ResourceTypeUnknownType,
+		MacAddress:       expandStringPtr(d.Get("mac_address")),
+		ResourceName:     nil,
+		ResourceIDs:      nil,
```
Comment (quantumsheep) [inline scaleway/data_source_ipam_ip.go:63]:
```suggestion
		PrivateNetworkID: expandStringPtr(d.Get("private_network_id")),
		ResourceID:       expandStringPtr(d.Get("resource_id")),
		ResourceType:     ipam.ResourceTypeUnknownType,
		MacAddress:       expandStringPtr(d.Get("mac_address")),
```

Simplify the code

File: scaleway/data_source_ipam_ip.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+				Optional: true,
+			},
+			"mac_address": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"type": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "IP Type (ipv4, ipv6)",
```
Comment (quantumsheep) [inline scaleway/data_source_ipam_ip.go:34]: Add string validation here instead of further in the code

**Review by quantumsheep — CHANGES_REQUESTED** [review]

File: scaleway/data_source_ipam_ip.go (hunk 2 above)
Comment (Codelax) [reply scaleway/data_source_ipam_ip.go:63]: This show clearly the missing filters, this is as efficient as a TODO that I will add to it

**Review by quantumsheep — APPROVED** [review]

**Pull request merged**